### PR TITLE
Autocomplete dialog with functional components

### LINKED ISSUE #1002: Refactor autocomplete dialog

### PR DESCRIPTION
**Pull request by ramirobg94**
Resolves #1002

### CODE REVIEW

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  className: PropTypes.string,
+  onItemSelect: PropTypes.func,
+  onClose: PropTypes.func,
+}
+
+AutocompleteDialog.defaultProps = {
+  items: [],
+  itemRenderer: null,
+  itemLabel: null, // required prop / string or function
+  itemKeyFunction: null, // required prop / string or function - Rename it to itemKey
```
Comment (minotogna) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:173]: as you can see from this comment, this should be renamed to itemKey and be oneOfType string or function

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        inputField.removeEventListener('keydown', onInputFieldKeyDown)
+      }
+
+      window.removeEventListener('click', onOutsideClick)
+    }
+  }, [inputField, onInputFieldKeyDown, onOutsideClick])
+
+  return (
+    <div ref={list} className={`autocomplete-list ${className}`} style={{ ...calculatePosition() }}>
+      {(items || []).map((item) => (
```
Comment (minotogna) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:143]: you dont need ```items || []```  there's a default value for items

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    },
+    [inputField, list, close]
+  )
+
+  const onInputFieldKeyDown = useCallback(
+    (event) => {
+      switch (event.keyCode) {
+        case KeyboardMap.Down:
+        case KeyboardMap.Tab:
+          if ((items || []).length > 0) {
```
Comment (minotogna) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:62]: you dont need items || []

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  const {
+    inputField,
+    sourceElement,
+    items,
+    itemRenderer: ItemRenderer,
+    itemLabel,
+    itemKeyFunction,
+    className,
+    onItemSelect: selectItem,
+    onClose: close,
```
Comment (minotogna) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:26]: is there a need to rename these ?

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+  useEffect(() => {
+    if (inputField) {
+      inputField.addEventListener('keydown', onInputFieldKeyDown)
+    }
+    window.addEventListener('click', onOutsideClick)
+
+    return () => {
+      if (inputField) {
+        inputField.removeEventListener('keydown', onInputFieldKeyDown)
```
Comment (minotogna) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:134]: use localState with actions

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    return {
+      top: top + height,
+      left,
+      width,
+    }
+  }, [sourceElement, inputField])
+
+  const focusItem = useCallback(
+    (index) => {
+      const itemEl = list.current && list.current.children && list.current.children[index]
```
Comment (SteRiccio) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:42]: I suppose that itemEl should never be null... so calling list.current.children[index] should be safe.

File: webapp/components/form/AutocompleteDialog/AutocompleteDialog.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    window.addEventListener('click', clickListener)
+
+    return () => {
+      if (inputField) {
+        inputField.removeEventListener('keydown', keyDownListener)
+      }
+
+      window.removeEventListener('click', clickListener)
+    }
+  }, [State.getInputField(state), list.current])
```
Comment (ramirobg94) [inline webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:59]: should i move that to some function called Actions.setListeners?

Comment (minotogna) [reply webapp/components/form/AutocompleteDialog/AutocompleteDialog.js:59]: no, you can add the entire useEffect to useAutocompleteDialog

**Pull request merged**